Convert SigstoreKeys to protobuf-specs type

codysoyland opened the pull request:
<!--
Thanks for opening a pull request! Please do not just delete this text.  The three fields below are mandatory.

Please remember to:
- This PR requires an issue. If it is a new feature, the issue should proceed the PR and will have allowed sufficent time for discussions to take place. Please use
issue tags such as "Closes #XYZ" or "Resolves sigstore/repo-name#XYZ".
  [documentation](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- ensure your commits are signed-off, as sigstore uses the [DCO](https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin) using `git commit -s`, or `git commit -s --amend` if you want to amend already existing commits
- lastly, ensure there are no merge commits!

Thank you :)
-->

#### Summary
<!--
 Explain the **motivation** for making this change. What existing problem does the pull request solve? How can reviewers test this PR?
-->

This PR converts [`SigstoreKeys`] into the [`TrustedRoot`](https://github.com/sigstore/protobuf-specs/blob/main/protos/sigstore_trustroot.proto#L101) type from `protobuf-specs`.

Additionally, this updates the TUF fetching logic to use `trusted_root.json` to fetch the trusted key material if it exists, and falls back to using the custom metadata and individual target files if that file is not included in given TUF repo.

This is the first step in modifying policy-controller to support bundle verification with [`sigstore-go`] using the OCI storage scheme detailed in [Sigstore Bundle as OCI Artifact].

Fixes https://github.com/sigstore/policy-controller/issues/1297

[`SigstoreKeys`]: https://github.com/sigstore/policy-controller/blob/main/pkg/apis/config/sigstore_keys.go#L84-L102
[`sigstore-go`]: https://github.com/sigstore/sigstore-go
[Sigstore Bundle as OCI Artifact]: https://github.com/sigstore/cosign/issues/3577

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to an administrator running private sigstore instances (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense.

-->

#### Documentation
<!--

Does this change require an update to documentation? How will users implement your new feature?

Please reference a PR within https://docs.sigstore.dev

-->

hectorj2f commented:
@codysoyland Let me know when it is ready for a review. Good work :)!

codysoyland commented:
> @codysoyland Let me know when it is ready for a review. Good work :)!

Thanks @hectorj2f! A few questions if you don't mind:

- The serialized test data in the reconciler tests (e.g. `marshalledEntry` and `marshalledEntryFromMirrorFS`) need to be regenerated any time I make a change to `SigstoreKeys`. I was thinking about replacing these with JSON files using `go:embed` and writing a script (added to `/hack`) to generate them, if you'd be open to that?
- I have the same question about the `validRepository`/`rootJSON` values in the tests (serialized TUF repos). You don't currently have a script to do this, do you? I see you [updated them somewhat recently](https://github.com/sigstore/policy-controller/pull/1155). If you have any notes, that would be helpful -- I'm assuming you used the go-tuf CLI?
- I don't want this patch to break anyone's existing K8s setup, but it does change the structure of the cached keys in the ConfigMap (`config-sigstore-keys`). The TrustRoot reconciler doesn't run automatically when upgrading the webhook, right? How would you ensure that the ConfigMap gets updated when they upgrade?
- To avoid breaking changes to the existing `v1alpha.SigstoreKeys` (k8s type), I was considering adding a new field to `TrustRootSpec` (e.g. `TrustedRootJSON:`) which would allow the user to just paste in a copy of a `trusted_root.json` (e.g. [this one from root-signing](https://github.com/sigstore/root-signing/blob/main/targets/trusted_root.json)). It seems simplest/safest to just accept it as a multiline string so that we don't need to maintain a yaml schema for it and the embedded `mediaType` could be used to deserialize it (within sigstore-go). Does this seem like a decent approach?

hectorj2f commented:
> have the same question about the validRepository/rootJSON values in the tests (serialized TUF repos). You don't currently have a script to do this, do you? I see you updated them somewhat recently. If you have any notes, that would be helpful -- I'm assuming you used the go-tuf CLI? Edit: Regarding the first two bullets, I finally read the comment stating that these values were pulled from a scaffolding deployment on a kind cluster! I did go ahead and move these values to separate files in a testdata subdir and added a minimal script to regenerate the marshalledEntry which saved me some time debugging since the diff output from TableTest is hard to read.

@codysoyland Would you mind creating a separate PR for this? So we can add these changes to main.

hectorj2f commented:
> The TrustRoot reconciler doesn't run automatically when upgrading the webhook, right? How would you ensure that the ConfigMap gets updated when they upgrade?

When you restart the webhook, it tries to reconcile all the TrustRoot resources.

If I understood you correctly. We have a watch on the configMap so every time it gets updated, we get the latest changes.

hectorj2f commented:
> For now, this PR will be constrained to only affecting internal data types and pulling the trusted root from TUF.

This sounds great. 
For any breaking change, we probably want to open one issue and discuss any decision with @vaikas and others.

codysoyland commented:
@hectorj2f I rebased this PR onto https://github.com/sigstore/policy-controller/pull/1325, added coverage for the fetching of `trusted_root.json` from TUF, and spent some time (along with @bdehamer) fixing all the tests/linter issues so I will mark it ready for review. After #1325 is merged, the diff will become a little smaller. Thanks in advance for your review!

hectorj2f commented on .github/workflows/build.yaml:
```diff
@@ -43,7 +43,7 @@ jobs:
 
       # will use the latest release available for ko
-      - uses: imjasonh/setup-ko@ace48d793556083a76f1e3e6068850c1f4a369aa # v0.6
+      - uses: ko-build/setup-ko@ace48d793556083a76f1e3e6068850c1f4a369aa # v0.6
```
I'd suggest to make these changes as part of a different PR

codysoyland replied:
Oddly, this failed to build in our private fork as the setup-ko action has moved -- not sure why it works here! I'll move them to a different PR.

codysoyland replied:
https://github.com/sigstore/policy-controller/pull/1328

hectorj2f commented on Makefile:
```diff
@@ -189,3 +189,6 @@ docs/generate-api:
 	    > docs/api-types/index-v1alpha1.md;
 
+.PHONY: generate-testdata
+generate-testdata:
```
It seems you need to rebase, I know we have merged this code.

vaikas reviewed (commented):
Nice! Not too bad after all, love it when a plan comes together 👍 Just need to do some rebasing as Hector says, but looks good overall! Thanks for handling my TODOs :)

codysoyland commented on pkg/apis/config/sigstore_keys.go:
```diff
@@ -133,56 +87,171 @@ func NewSigstoreKeysFromConfigMap(config *corev1.ConfigMap) (*SigstoreKeysMap, e
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to unmarshal public key: %w", err)
+	}
+	var keyDetails pbcommon.PublicKeyDetails
+	switch k := pk.(type) {
+	case *ecdsa.PublicKey:
+		switch k.Curve {
+		case elliptic.P256():
+			keyDetails = pbcommon.PublicKeyDetails_PKIX_ECDSA_P256_SHA_256
+		case elliptic.P384():
+			keyDetails = pbcommon.PublicKeyDetails_PKIX_ECDSA_P384_SHA_384
+		case elliptic.P521():
+			keyDetails = pbcommon.PublicKeyDetails_PKIX_ECDSA_P521_SHA_512
+		default:
+			keyDetails = pbcommon.PublicKeyDetails_PUBLIC_KEY_DETAILS_UNSPECIFIED
+		}
+	case *rsa.PublicKey:
+		switch k.Size() {
+		case 2048:
+			keyDetails = pbcommon.PublicKeyDetails_PKIX_RSA_PSS_2048_SHA256
+		case 3072:
+			keyDetails = pbcommon.PublicKeyDetails_PKIX_RSA_PSS_3072_SHA256
+		case 4096:
+			keyDetails = pbcommon.PublicKeyDetails_PKIX_RSA_PSS_4096_SHA256
+		default:
+			keyDetails = pbcommon.PublicKeyDetails_PUBLIC_KEY_DETAILS_UNSPECIFIED
+		}
+	default:
+		keyDetails = pbcommon.PublicKeyDetails_PUBLIC_KEY_DETAILS_UNSPECIFIED
+	}
```
I need to spend a few moments to confirm these assumptions with regard to key type. @kommendorkapten, if you have a moment to weigh in, that would be helpful! This code path is exercised when a transparency log public key is provided in a custom resource (PEM encoded), or when a TUF repo does not have a `trusted_root.json`. So it does not apply to the Sigstore Public Good instance, which provides a `trusted_root.json`. For private transparency logs, we need to assume the key type to fill in the values required by the TrustedRoot.

kommendorkapten replied:
I think this code is solid. It uses Go std `ParsePKIXPublicKey`, and the created public key should so have the correct parameters.

The only thing that possibly could break is for RSA is the intended signature scheme is PKCS1#v1.5, but without any OOB information, we can't tell which signature scheme to use.

Do you know where in the code we load the verifier for RSA key? I can also take a look, but there are two options in sigstore/sigstore:
PKCS#1v1.5: https://github.com/sigstore/sigstore/blob/main/pkg/signature/rsapkcs1v15.go#L121
PSS: https://github.com/sigstore/sigstore/blob/main/pkg/signature/rsapss.go#L147

kommendorkapten replied:
Ok, so I think we should change things here.

There are two places I found where a verifier is loaded 
* https://github.com/sigstore/policy-controller/blob/fc2800a5e9cf28f75c63e5ac6f50e2e468f6b5a2/pkg/webhook/validation.go#L38 
* https://github.com/sigstore/policy-controller/blob/fc2800a5e9cf28f75c63e5ac6f50e2e468f6b5a2/pkg/webhook/validator.go#L839

Both of them calls `sigstore/pkg/signature/verifier::LoadVerifier`. That functions will assume PKCS#1v1.5 padding. To use PSS padding, you need to explicitly load that via `signature.LoadRSAPSSVerifier`, there are no such references in this code base.

So in the code above, you should load:
* `PKIX_RSA_PKCS1V15_2048_SHA256`
* `PKIX_RSA_PKCS1V15_3072_SHA256`
* `PKIX_RSA_PKCS1V15_4096_SHA256`

IMHO this should be clarified in the docs too, that when using RSA keys, PKCS#1v1.5 padding is only supported.

codysoyland replied:
Thanks @kommendorkapten! I will file an issue for adding the ability use a protojson trusted root in the TrustRoot CRD which will make this section of code obsolete, and by the way, these key details are not actually used anywhere in Policy Controller at the moment -- They will only be used when we add the ability to use Sigstore Bundles as OCI artifacts. So this is not a regression and the trusted root support should be complete before we add any support for bundles.

kommendorkapten commented on pkg/apis/config/sigstore_keys.go:
```diff
@@ -133,56 +87,171 @@ func NewSigstoreKeysFromConfigMap(config *corev1.ConfigMap) (*SigstoreKeysMap, e
+	case *rsa.PublicKey:
+		switch k.Size() {
+		case 2048:
+			keyDetails = pbcommon.PublicKeyDetails_PKIX_RSA_PSS_2048_SHA256
```
```suggestion
			keyDetails = pbcommon.PublicKeyDetails_PKIX_RSA_PKCS1V15_2048_SHA256
```

kommendorkapten commented on pkg/apis/config/sigstore_keys.go:
```diff
@@ -133,56 +87,171 @@ func NewSigstoreKeysFromConfigMap(config *corev1.ConfigMap) (*SigstoreKeysMap, e
+		case 2048:
+			keyDetails = pbcommon.PublicKeyDetails_PKIX_RSA_PSS_2048_SHA256
+		case 3072:
+			keyDetails = pbcommon.PublicKeyDetails_PKIX_RSA_PSS_3072_SHA256
```
```suggestion
			keyDetails = pbcommon.PublicKeyDetails_PKIX_RSA_PKCS1V15_3072_SHA256
```

kommendorkapten commented on pkg/apis/config/sigstore_keys.go:
```diff
@@ -133,56 +87,171 @@ func NewSigstoreKeysFromConfigMap(config *corev1.ConfigMap) (*SigstoreKeysMap, e
+		case 3072:
+			keyDetails = pbcommon.PublicKeyDetails_PKIX_RSA_PSS_3072_SHA256
+		case 4096:
+			keyDetails = pbcommon.PublicKeyDetails_PKIX_RSA_PSS_4096_SHA256
```
```suggestion
			keyDetails = pbcommon.PublicKeyDetails_PKIX_RSA_PKCS1V15_4096_SHA256
```

codysoyland commented:
I rebased again and also removed unnecessary go.mod updates, which in turn removed the commits related to run-running codegen, so I think this PR is in a good state to merge now. Would love to have this reviewed again @hectorj2f @vaikas 🙏🏻  Thanks!

hectorj2f commented on hack/gentestdata/gentestdata.go:
```diff
@@ -1,10 +1,10 @@
-// Copyright 2024 The Sigstore Authors
+// Copyright 2022 The Sigstore Authors.
```
2024 ?

hectorj2f reviewed (dismissed):
LGTM overall

vaikas commented on pkg/apis/config/sigstore_keys.go:
```diff
@@ -133,56 +87,171 @@ func NewSigstoreKeysFromConfigMap(config *corev1.ConfigMap) (*SigstoreKeysMap, e
+func ConvertTransparencyLogInstance(source v1alpha1.TransparencyLogInstance) *pbtrustroot.TransparencyLogInstance {
+	pbpk, pk, err := DeserializePublicKey(source.PublicKey)
+	if err != nil {
+		return nil // TODO: log error? Add return error?
```
Yeah, I think we should return the error instead of silently eating it. Here and in L151, since we return it from DeserializePublicKey below.

vaikas reviewed (approved):
Thanks!
